File: sample-Android/Suprdemo/supr/src/main/java/com/impraise/supr/data/DataResult.kt
package com.impraise.supr.data

/**
 * Created by guilhermebranco on 1/5/18.
 */
open class DataResult<out T> protected constructor(val state: DataResultState, val data: T?, val error: Throwable?) {

    companion object {

        fun <T> success(data: T): DataResult<T> {
            return DataResult(DataResultState.Success, data, null)
        }

        fun <T> error(error: Throwable, data: T?): DataResult<T> {
            return DataResult(DataResultState.Error, data, error)
        }
    }
}

File: sample-iOS/Supr-demo/Supr-demo/Supr/Data/DataResult.swift
//
//  DataResult.swift
//  Supr-demo
//
//  Created by Arnaud Camus on 2/22/18.
//  Copyright © 2018 Impraise. All rights reserved.
//

class DataResult<T> {
    let state: DataResultState
    let data: T?
    let error: Error?
    
    init(_ state: DataResultState, data: T? = nil, error: Error? = nil) {
        self.state = state
        self.data = data
        self.error = error
    }
    
    static func success(data: T) -> DataResult<T> {
        return DataResult(DataResultState.success, data: data)
    }
    
    static func error(error: Error, data: T?) -> DataResult<T> {
        return DataResult(DataResultState.error, data: data, error: error)
    }
}
